src/debug_wait.c

```diff
@@ -102,81 +210,32 @@ ts_debug_waitpoint_wait(DebugWait *waitpoint, bool blocking)
 	}
 	Assert(lock_acquire_result == LOCKACQUIRE_OK);
 
-	lock_release_result = LockRelease(&waitpoint->tag, ShareLock, true);
+	lock_release_result = LockRelease(&point.tag, ShareLock, true);
 	Assert(lock_release_result);
 
-	ereport(DEBUG3, (errmsg("proceeding after waitpoint '%s'", waitpoint->tagname)));
-}
-
-static void
-debug_waitpoint_enable(DebugWait *waitpoint)
-{
-	ereport(DEBUG1, (errmsg("enabling waitpoint \"%s\"", waitpoint->tagname)));
-	if (LockAcquire(&waitpoint->tag, ExclusiveLock, true, true) == LOCKACQUIRE_NOT_AVAIL)
-		ereport(NOTICE, (errmsg("debug waitpoint \"%s\" already enabled", waitpoint->tagname)));
-}
-
-static void
-debug_waitpoint_release(DebugWait *waitpoint)
-{
-	ereport(DEBUG1, (errmsg("releasing waitpoint \"%s\"", waitpoint->tagname)));
-	if (!LockRelease(&waitpoint->tag, ExclusiveLock, true))
-		elog(ERROR, "cannot release waitpoint \"%s\"", waitpoint->tagname);
+	ereport(DEBUG3, (errmsg("proceeding after debug point '%s'", point.name)));
 }
 
 /*
- * Enable a waitpoint to block when being reached.
- *
- * This function will always succeed since we will not lock the waitpoint if
- * it is already locked. A notice will be printed if the waitpoint is already
- * enabled.
+ * Produce an error in case if the debug point is enabled.
```

Comment:
definitely need more comments here to explain the logic of how the error injection works. Something like:

```
The idea is to enable the debug point separately first which acquires a ShareLock on this tag. With the debug point enabled, this function when invoked will not get the exclusive lock and will be able to raise the error as desired.
```

Reply:
I can extend it, but more expanded comment is given in the header

Reply:
I would prefer a slightly different API, like enable_debug_waitpoint/disable_debug_waitpoint/...; enable_debug_error_injection/disable_debug_error_injection/... Currently, it's slightly messed up and disclosures the implementation details. I don't think that's a major problem though and we can refactor it in another patch.